server/src/destination_device_validator.rs

```diff
@@ -0,0 +1,98 @@
+use anyhow::{bail, Result};
+use std::iter::zip;
+
+use common::web_api::{SignalMessage, SignalMessages};
+
+use crate::account::Account;
+
+pub struct DestinationDeviceValidator;
+
+impl DestinationDeviceValidator {
+    pub fn validate_registration_id_from_messages(
+        account: &Account,
+        messages: &[SignalMessage],
+        use_phone_number_identity: bool,
+    ) -> Result<()> {
+        let destination_device_ids: Vec<u8> = messages
+            .iter()
+            .map(|message| message.destination_device_id)
+            .collect();
+        let destination_registration_ids: Vec<u32> = messages
+            .iter()
+            .map(|message| message.destination_registration_id)
+            .collect();
+        let device_id_and_registration_id: Vec<(u8, u32)> =
+            zip(destination_device_ids, destination_registration_ids).collect();
+        Self::validate_registration_id_from_id_tuple(
+            account,
+            &device_id_and_registration_id,
+            use_phone_number_identity,
+        )
+    }
+
+    pub fn validate_registration_id_from_id_tuple(
+        account: &Account,
+        device_ids_and_registration_ids: &[(u8, u32)],
+        use_phone_number_identity: bool,
+    ) -> Result<()> {
+        let stale_devices: Vec<u8> = device_ids_and_registration_ids
+            .iter()
+            .filter(|device_id_and_registration_id| device_id_and_registration_id.1 > 0)
+            .filter(|device_id_and_registration_id| {
+                let device_id = device_id_and_registration_id.0;
+                let registration_id = device_id_and_registration_id.1;
+                let registration_id_matches: bool = if let Some(device) = account
+                    .devices()
+                    .iter()
+                    .find(|device| (u32::from(device.device_id()) as u8) == device_id)
+                {
+                    registration_id
+                        == if use_phone_number_identity {
+                            device.pni_registration_id()
+                        } else {
+                            device.registration_id()
+                        }
```

Comment:
Maybe this could be internal logic within the struct, so when calling device.registration_id(), it figures out which to use

Reply:
Will just keep to match Signal Server logic